### PR TITLE
Fix for #2877 - add nonanalyzed (i.e. keyword) 'name' field to package search

### PR DESCRIPTION
**Pull request by selotape**
To whom it may concern,

I fixed the bug to the best of my understanding, but I still have a few issues. I'm "prematurely" opening this PR with the purpose of receiving some review, feedback, and help.

See some code comments about the fix inline.

About my status:

1. I managed to set up and run warehouse in a container (😄)
2. I couldn't get the tests to execute :\. When I run "make tests" I get the following error:
```
    env: ‘bin/tests’: No such file or directory
    Makefile:99: recipe for target 'tests' failed
```
3. I couldn't replicate the bug on my PC, hence I couldn't make sure the bug is solved. Finding "bad" examples is hard even in production, and the test ES probably has much fewer packages. The bug can probably be fabricated by adding carefully named mock packages to the index, but I've no idea how to do that.

4. The "tests" I did run were making sure that 1) search still works, 2) `make initdb` still succeeds and 3) the ES search query contains the new search field, `name.keyword`.

Any help and guidance would be appreciated.
Thanks!

### CODE REVIEW

File: warehouse/packaging/search.py [hunk 1]
```diff
@@ -33,7 +33,7 @@
 @doc_type
 class Project(Document):
 
-    name = Text()
+    name = Text(fields={'keyword': Keyword()})
```
Comment (selotape) [inline warehouse/packaging/search.py]: Here is where I add the inner field `name.keyword`, with the keyword type.
This is following @HonzaKral s suggestion.

File: warehouse/search/queries.py [hunk 2]
```diff
@@ -10,26 +10,21 @@
...
     "summary": 5,
+    "author": 1,
+    "author_email": 1,
+    "download_url": 1,
+    "home_page": 1,
+    "license": 1,
+    "maintainer": 1,
+    "maintainer_email": 1,
+    "platform": 1,
 }
```
Comment (selotape) [inline warehouse/search/queries.py]: This is a small refactoring - I removed the duplicate definition of search fields. Now you only need to state search fields once and you must set a boost for it.

File: warehouse/views.py [hunk 3]
```diff
@@ -413,8 +413,8 @@ def form_query(query_type, query):
     Returns a multi match query
     """
     fields = [
-        field + "^" + str(SEARCH_BOOSTS[field]) if field in SEARCH_BOOSTS else field
-        for field in SEARCH_FIELDS
+        field + "^" + str(boost) if boost != 1 else field
+        for field, boost in SEARCH_BOOSTS.items()
     ]
```
Comment (selotape) [inline warehouse/views.py]: This is the same refactoring as above

File: warehouse/packaging/search.py (hunk 1 above)
Comment (selotape) [reply warehouse/packaging/search.py]: FYI - I saw no need to add a "deserialization" of `name.keyword` in the `from_db` class method. So I didn't do it :)

**Comment by di** [comment]
> I couldn't get the tests to execute :. When I run "make tests" I get the following error:

Is there more to this error? I would imagine there would be a stack trace or something similar above what you included.

> I couldn't replicate the bug on my PC, hence I couldn't make sure the bug is solved. Finding "bad" examples is hard even in production, and the test ES probably has much fewer packages. The bug can probably be fabricated by adding carefully named mock packages to the index, but I've no idea how to do that.

You can upload new distributions directly to your development instance of Warehouse with `twine`, see http://warehouse.readthedocs.io/development/reviewing-patches/#testing-with-twine for instructions. I like to use https://github.com/pypa/sampleproject for this purpose, and modify it as needed.

You should be able to reproduce this by adding two new packages to your development instance.

You'll need to run `make reindex` afterwards to rebuild the elasticsearch index to include the new packages.

**Comment by ewdurbin** [comment]
> You'll need to run make reindex afterwards to rebuild the elasticsearch index to include the new packages.

With #4328 that _shouldn't_ be the case as long as the workers are running!

**Comment by ewdurbin** [comment]
@selotape it sounds like you're using a bit of a non-standard environment to develop warehouse.

Could you help us by describing how you're running warehouse? Our expected development environment is documented at https://warehouse.readthedocs.io/development/getting-started/#detailed-installation-instructions, if you're diverging from that too much it can be difficult for contributors to assist you.

If you're running into snags with the documented setup, that's a bug and we encourage you to file an issue!

**Comment by di** [comment]
Oh yeah, forgot that we fixed that!

**Comment by selotape** [comment]
@ewdurbin , 

about my setup, I did follow the "detailed installation instructions", but I'm not familiar with docker so I may have gotten some things wrong.

My main OS is windows 10 so I installed ubuntu on the WSL. I installed docker CE in windows and docker and docker-compose on the Ubuntu. I also followed the "quirks" section. After all this I ran `make serve` and `make initdb` and things seemed to run well, including all the sanity tests (except `make tests`...).

Is there anything specific I can check and show you to make sure I'm set up properly?



And about the tests, the following is all I get when I run `make tests` from the project root - 
image


I find the "no such file" error strange, because the `bin/tests` file does exist - 
image

...maybe it's somehow related to the fact that I'm user `root` on the Ubuntu... and the docker can't see the tests...?
Any other suggestions for how to run the tests?

Thanks!

**Comment by selotape** [comment]
btw, `make lint` also doesn't work for me. It looks like a python3.6 issue - 
image

Is the python3.6 supposed to be installed on the container or on the Ubuntu?

File: warehouse/packaging/search.py (hunk 1 above)
Comment (HonzaKral) [reply warehouse/packaging/search.py]: That's not needed - the `name.keyword` does not acctually appear on the document, it is just an alternative index that can be used for searching and/or aggregations/sorting.

**Comment by brainwane** [comment]
@selotape I'm so sorry for the delay in response! Short answer: [we're in IRC most weekdays around 10am-5pm US Eastern Time](https://webchat.freenode.net/?channels=%23pypa-dev,pypa), come talk with us?

Longer answer: [From December 2017 till the end of April 2018, PyPI had funding to get the new site up and running and perform the switchover.](https://wiki.python.org/psf/PackagingWG) Then the grant money ran out. Then the Packaging Working Group, seeking [donations](http://donate.pypi.org/) and further grants to fund more work, got [some new funding](https://pyfound.blogspot.com/2018/12/upcoming-pypi-improvements-for-2019.html), and we're getting our momentum back on Warehouse work and being able to spend more time helping contributors. So if you come talk with us in IRC sometime soon we should be able to help you get your setup fixed! Or if you'll be at PyCon North America in May, [talk with us during the sprints](https://wiki.python.org/psf/PackagingSprints)!

Thanks and sorry again.

**Pull request closed without merging**